Fix jump_backwards behaviour when jumplist is at capacity

thomasschafer opened the pull request:
This PR resolves some issues with the jumplist when the capacity is reached. In particular:
- Removes `self.jumps.truncate` from inside `JumpList::push` - `truncate` removes entries from the end, but of course we want to remove entries from the beginning, as those are the oldest entries. The jumplist is shortened using `pop_front` later on before pushing any new elements to the end, so we don't need to truncate it in advance.
- Decrements the `current` position in the jumplist if elements are removed from the front: if an element in the jumplist shifts backwards then the index pointing to that element should shift backwards with it.
- Checks whether the previous position in the jumplist is the position that the cursor is currently on: if so, move another step back in the jumplist, as otherwise the cursor stays in the same place and nothing happens. Previously the `truncate` (mentioned above) was a hacky resolution to this issue, but now that the truncation is removed we can fix the issue properly.

Fixes https://github.com/helix-editor/helix/issues/10967.

thomasschafer commented:
Realised that tests are failing - away from my laptop but will fix when I can

the-mikedavis reviewed (changes requested):
Thanks for opening this up! I've seen this bug before but couldn't figure out how to reproduce the behavior.

As you note in a comment in `backwards`, the issue is that we're using `current` as a sort of pointer and when we pop items from the front, the index of our "current" item changes. The fix for this I believe is smaller than what you have at the moment here: in `backward` we can make `current` `mut` and subtract the number of items we will remove in `push` (`self.jumps.len() - JUMP_LIST_CAPACITY + 1`) in the `if self.current == self.jumps.len()` block.

thomasschafer commented:
> Thanks for opening this up! I've seen this bug before but couldn't figure out how to reproduce the behavior.
> 
> As you note in a comment in `backwards`, the issue is that we're using `current` as a sort of pointer and when we pop items from the front, the index of our "current" item changes. The fix for this I believe is smaller than what you have at the moment here: in `backward` we can make `current` `mut` and subtract the number of items we will remove in `push` (`self.jumps.len() - JUMP_LIST_CAPACITY + 1`) in the `if self.current == self.jumps.len()` block.

Thank you for taking a look!

Is it not preferable to update `current` in `push`, rather than `backwards`, as that ensures that any other uses of `push` also ensure that `current` is valid? I can't see why we'd want to use `push` while allowing elements to be popped from the front of the jump list and not also keep `current` in sync.

the-mikedavis commented:
`push` can modify `self.current` as it currently does on master but it can't automatically adjust the `current` binding in `backward`. I mean the `current` binding in the `if let Some` here: https://github.com/helix-editor/helix/blob/dbacaaddcaa4a9dd39c73d0102d03ea7c24ca647/helix-view/src/view.rs#L66

The current changes to `push` actually do not change the behavior. Decrementing `self.current` while popping elements isn't necessary because `self.current` isn't read anywhere between the popping and assignment of `self.current = self.jumps.len()`. And moving that assignment to the end of the block also doesn't have an effect because `self.jumps.truncate(self.current)` implies that `self.current` then equals `self.jumps.len()`

thomasschafer commented:
> `push` can modify `self.current` as it currently does on master but it can't automatically adjust the `current` binding in `backward`. I mean the `current` binding in the `if let Some` here:
> 
> https://github.com/helix-editor/helix/blob/dbacaaddcaa4a9dd39c73d0102d03ea7c24ca647/helix-view/src/view.rs#L66
> 
> The current changes to `push` actually do not change the behavior. Decrementing `self.current` while popping elements isn't necessary because `self.current` isn't read anywhere between the popping and assignment of `self.current = self.jumps.len()`. And moving that assignment to the end of the block also doesn't have an effect because `self.jumps.truncate(self.current)` implies that `self.current` then equals `self.jumps.len()`

Ah I see, you're right - thanks for explaining. I've removed the changes from `push`. I've also updated `backward` in line with your comment [here](https://github.com/helix-editor/helix/pull/10968/#pullrequestreview-2120626727) with a slight modification: I felt uneasy about performing the subtraction in `backward` to try and determine how many elements had been popped from the front in `push`, so I've updated `push` to return the number of elements removed from the front and `backward` now subtracts those from `current`, as per your comment. Hopefully this is okay but happy to make changes if not!

the-mikedavis commented on helix-view/src/view.rs:
```diff
@@ -38,18 +38,22 @@ impl JumpList {
     }
 
-    pub fn push(&mut self, jump: Jump) {
+    pub fn push(&mut self, jump: Jump) -> usize {
```
Returning the number from `push` is ok but I don't want to let this leak out into the public interface. Let's rename this `push` to `push_impl` and make it private and then introduce a new `pub fn push` that has no return and just calls `self.push_impl(jump);`

thomasschafer replied:
Sounds good, done

the-mikedavis commented on helix-view/src/view.rs:
```diff
@@ -38,18 +38,22 @@ impl JumpList {
+            // If the jumplist is full, drop the oldest items.
             while self.jumps.len() >= JUMP_LIST_CAPACITY {
-                self.jumps.pop_front();
+                if self.jumps.pop_front().is_some() {
```
This is guaranteed to be `Some` because `jumps.len()` is greater than the capacity which is `> 0`.

thomasschafer replied:
Updated

the-mikedavis commented on helix-view/src/view.rs:
```diff
@@ -63,13 +67,23 @@ impl JumpList {
             }
             self.current = current;
-            self.jumps.get(self.current)
+            self.jumps.get(self.current).and_then(|prev_jump| {
+                // If a jumplist position is saved and then `jump_backward` is immediately invoked
+                // without moving the cursor, jump back one more step to prevent a no-op
+                if prev_jump == &cur_jump {
+                    self.current = self.current.saturating_sub(1);
+                    self.jumps.get(self.current)
+                } else {
+                    Some(prev_jump)
+                }
+            })
```
Changing the behavior here sounds good to me but the `and_then` can be a little tough to read. Let's use the `?` operator to clean it up and also avoid making another clone of the selection to check equality (i.e. not binding `cur_jump` above):

```rust
// Avoid jumping to the current location.
let jump @ (doc_id, selection) = self.jumps.get(self.current)?;
if doc.id() == *doc_id && doc.selection(view_id) == selection {
    self.current = self.current.checked_sub(1)?;
    self.jumps.get(self.current)
} else {
    Some(jump)
}
```

thomasschafer replied:
Looks good - updated
